[Merged by Bors] - feat: add the portmanteau implication from Borel set limit condition to closed set limsup condition

kkytola opened the pull request:
Add one implication of portmanteau theorem: a Borel set limit condition implies a closed set limsup condition.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [x] depends on: #6225
- [x] depends on: #6194
-->
- [x] depends on: #6225
- [x] depends on: #6194

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

RemyDegenne commented on Mathlib/MeasureTheory/Measure/Portmanteau.lean:
```diff
@@ -456,9 +452,60 @@ theorem exists_null_frontiers_thickening (μ : Measure Ω) [SigmaFinite μ] (s :
+Assuming that for all Borel sets `E` whose boundary `∂E` carries no probability mass under a
+candidate limit probability measure `μ` we have convergence of the measures `μs i E` to `μ E`,
+then for all open sets `G` we have the limsup condition `μ G ≤ liminf (μs i G)`. -/
+lemma le_liminf_measure_open_of_forall_tendsto_measure
+    {Ω ι : Type _} {L : Filter ι} [NeBot L]
+    [MeasurableSpace Ω] [PseudoEMetricSpace Ω] [OpensMeasurableSpace Ω]
+    {μ : Measure Ω} [IsProbabilityMeasure μ] {μs : ι → Measure Ω} [∀ i, IsProbabilityMeasure (μs i)]
+    (h : ∀ {E : Set Ω}, MeasurableSet E → (μ : Measure Ω) (frontier E) = 0 →
+            Tendsto (fun i ↦ (μs i : Measure Ω) E) L (𝓝 ((μ : Measure Ω) E)))
+    (G : Set Ω) (G_open : IsOpen G) :
+    (μ : Measure Ω) G ≤ L.liminf (fun i ↦ (μs i : Measure Ω) G) := by
```
You can simply write `μs i E` instead of `(μs i : Measure Ω) E`.
```suggestion
lemma le_liminf_measure_open_of_forall_tendsto_measure
    {Ω ι : Type _} {L : Filter ι} [NeBot L]
    [MeasurableSpace Ω] [PseudoEMetricSpace Ω] [OpensMeasurableSpace Ω]
    {μ : Measure Ω} [IsProbabilityMeasure μ] {μs : ι → Measure Ω} [∀ i, IsProbabilityMeasure (μs i)]
    (h : ∀ {E}, MeasurableSet E → μ (frontier E) = 0 → Tendsto (fun i ↦ μs i E) L (𝓝 (μ E)))
    (G : Set Ω) (G_open : IsOpen G) :
    μ G ≤ L.liminf (fun i ↦ μs i G) := by
```

RemyDegenne commented on Mathlib/MeasureTheory/Measure/Portmanteau.lean:
```diff
@@ -456,9 +452,60 @@ theorem exists_null_frontiers_thickening (μ : Measure Ω) [SigmaFinite μ] (s :
+  let rs := Classical.choose ex
+  have rs_lim : Tendsto rs atTop (𝓝 0) := (Classical.choose_spec ex).1
+  have rs_pos : ∀ n, 0 < rs n := fun n ↦ ((Classical.choose_spec ex).2 n).1
+  have rs_null : ∀ n, (μ : Measure Ω) (frontier (Metric.thickening (rs n) F)) = 0 :=
```
```suggestion
  have rs_null : ∀ n, μ (frontier (Metric.thickening (rs n) F)) = 0 :=
```

RemyDegenne commented on Mathlib/MeasureTheory/Measure/Portmanteau.lean:
```diff
@@ -456,9 +452,60 @@ theorem exists_null_frontiers_thickening (μ : Measure Ω) [SigmaFinite μ] (s :
+  intros ε ε_pos μF_finite
+  have keyB := tendsto_measure_cthickening_of_isClosed (μ := μ) (s := F)
+                ⟨1, ⟨by simp only [gt_iff_lt, zero_lt_one], measure_ne_top _ _⟩⟩ F_closed
+  have nhd : Iio ((μ : Measure Ω) F + ε) ∈ 𝓝 ((μ : Measure Ω) F) := by
```
```suggestion
  have nhd : Iio (μ F + ε) ∈ 𝓝 (μ F) := by
```

RemyDegenne commented on Mathlib/MeasureTheory/Measure/Portmanteau.lean:
```diff
@@ -456,9 +452,60 @@ theorem exists_null_frontiers_thickening (μ : Measure Ω) [SigmaFinite μ] (s :
+  simp only [mem_map, mem_atTop_sets, ge_iff_le, mem_preimage, mem_Iio] at rs_lim
+  obtain ⟨m, hm⟩ := rs_lim
+  have aux' := fun i ↦ measure_mono (μ := μs i) (Metric.self_subset_thickening (rs_pos m) F)
+  have aux : (fun i ↦ ((μs i : Measure Ω) F))
+              ≤ᶠ[L] (fun i ↦ (μs i : Measure Ω) (Metric.thickening (rs m) F)) := by
+    exact eventually_of_forall aux'
```
```suggestion
  have aux : (fun i ↦ (μs i F)) ≤ᶠ[L] (fun i ↦ μs i (Metric.thickening (rs m) F)) :=
    eventually_of_forall aux'
```

RemyDegenne commented on Mathlib/MeasureTheory/Measure/Portmanteau.lean:
```diff
@@ -456,9 +452,60 @@ theorem exists_null_frontiers_thickening (μ : Measure Ω) [SigmaFinite μ] (s :
+    {Ω ι : Type _} {L : Filter ι} [NeBot L]
+    [MeasurableSpace Ω] [PseudoEMetricSpace Ω] [OpensMeasurableSpace Ω]
+    {μ : Measure Ω} [IsProbabilityMeasure μ] {μs : ι → Measure Ω} [∀ i, IsProbabilityMeasure (μs i)]
+    (h : ∀ {E : Set Ω}, MeasurableSet E → (μ : Measure Ω) (frontier E) = 0 →
+            Tendsto (fun i ↦ (μs i : Measure Ω) E) L (𝓝 ((μ : Measure Ω) E)))
```
```suggestion
    (h : ∀ {E}, MeasurableSet E → μ (frontier E) = 0 → Tendsto (fun i ↦ μs i E) L (𝓝 (μ E)))
```

RemyDegenne commented on Mathlib/MeasureTheory/Measure/Portmanteau.lean:
```diff
@@ -456,9 +452,60 @@ theorem exists_null_frontiers_thickening (μ : Measure Ω) [SigmaFinite μ] (s :
+                ⟨1, ⟨by simp only [gt_iff_lt, zero_lt_one], measure_ne_top _ _⟩⟩ F_closed
+  have nhd : Iio ((μ : Measure Ω) F + ε) ∈ 𝓝 ((μ : Measure Ω) F) := by
+    apply Iio_mem_nhds
+    simpa only [add_zero] using ENNReal.add_lt_add_left μF_finite.ne (ENNReal.coe_pos.mpr ε_pos)
```
```suggestion
    exact ENNReal.lt_add_right μF_finite.ne (ENNReal.coe_pos.mpr ε_pos).ne'
```

RemyDegenne reviewed (commented):
This looks good. Can you make sure that the file docstring is up to date with these new changes?

kkytola commented:
> This looks good. Can you make sure that the file docstring is up to date with these new changes?

Thank you for reminding! I updated the module docstring and rewrote the module docstring and section docstrings to be more systematic.

RemyDegenne commented:
Thanks!
bors r+
